_overviews/overview/a-taste-of-scala.md

```diff
@@ -0,0 +1,1329 @@
+
+// no parens
+while
+  x < 3
+do
+  println(x)
+  x += 1
+```
+
+### Create your own control structures!
+
+In Scala you can also create code that works just like a control structure. You can learn more about this in the Control Structures chapter in the Reference documentation.
+<!--
+NOTE: my old examples are `whilst` and `doubleIf`; `using` is a good example.
+-->
+
+
+
+## Data Modeling
+
+Scala supports both functional programming (FP) and object-oriented programming (OOP), as well as a fusion of the two paradigms. This section provides a quick overview of data modeling in OOP and FP.
+
+
+### OOP data modeling (traits and classes)
+
+When writing code in an OOP style, your two main tools will be *traits* and *classes*.
+
+#### Traits
+<!--
+- traits as interface
+- implemented methods
+- class extends traits
+- traits as mixins
+-->
+
+Traits are like interfaces in other languages, but they can also contain implemented methods. They provide a great way for you to organize behaviors into small, modular units. When you want to create concrete implementations of attributes and behaviors, classes and objects can extend traits, mixing in as many traits as needed to achieve the desired behavior.
+
+Here are three traits that define well-organized and modular behaviors for animals like dogs and cats:
+
+```scala
+trait Speaker:
+  def speak(): String  // has no body, so it’s abstract
+
+trait TailWagger:
+  def startTail(): Unit = println("tail is wagging")
+  def stopTail(): Unit = println("tail is stopped")
+
+trait Runner:
+  def startRunning(): Unit = println("I’m running")
+  def stopRunning(): Unit = println("Stopped running")
+```
+
+Given those traits, here’s a `Dog` class that extends all of those traits while providing a behavior for the abstract `speak` method:
+
+```scala
+class Dog(name: String) extends Speaker with TailWagger with Runner:
+  def speak(): String = "Woof!"
+```
+
+Notice how the class extends the traits with the `extends` and `with` keywords.
+
+Similarly, here’s a `Cat` class that implements those same traits while also overriding two of the concrete methods it inherits:
+
+```scala
+class Cat(name: String) extends Speaker with TailWagger with Runner:
+  def speak(): String = "Meow"
+  override def startRunning(): Unit = println("Yeah ... I don’t run")
+  override def stopRunning(): Unit = println("No need to stop")
+```
+
+These examples show how those classes are used:
+
+```scala
+val d = Dog("Rover")
+d.speak()               // prints "Woof!"
+
+val c = Cat("Morris")
+c.speak()               // "Meow"
+c.startRunning()        // "Yeah ... I don’t run"
+c.stopRunning()         // "No need to stop"
+```
+
+<!-- TODO: links  -->
+If that code makes sense — great, you’re comfortable with traits as interfaces. If not, don’t worry, they’re explained in more detail in the Data Modeling sections of this Overview and the Reference documentation.
+
+<!-- TODO: show traits that take parameters here? -->
+
+
+#### Classes
```

Comment:
Note that we have used class definitions above without introducing the concept of class.

Reply:
Yes, I have a little bit of a “chicken and the egg” thing going on. I’ll try to resolve this.